lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -2416,6 +2427,96 @@ selectSeqStatePendingIxs wid =
   where
     fromRes = fmap (W.Index . seqStatePendingIxIndex . entityVal)
 
+instance
+    ( PersistPublicKey (k 'AccountK)
+    , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+    , PaymentAddress n k
+    , SoftDerivation k
+    , WalletKey k
+    , Typeable n
+    ) => PersistState (Shared.SharedState n k) where
+    insertState (wid, sl) st = case st of
+        Shared.PendingSharedState prefix accXPub pTemplate dTemplateM g -> do
+            insertSharedState accXPub g pTemplate dTemplateM prefix
+            insertCosigner (cosigners pTemplate) Payment
+            when (isJust dTemplateM) $
+                insertCosigner (fromJust $ cosigners <$> dTemplateM) Delegation
```

Comment:
This is probably why the state machine tests fail.
Suppose two arbitrary checkpoints are added, and the second contains fewer cosigners than the first. There is no deletion of rows from the `cosigner_key` table, only insertion. Next time the checkpoint is read, there will be more cosigners than expected.

Reply:
Indeed. `insertState` should make sure that a given state can be read back. So, it's important here that existing cosigners are purged before calling `insertCosigner`.

Reply:
thanks for the suggestions. I tried to apply purging, but, unfortunately still not passing https://github.com/input-output-hk/cardano-wallet/pull/2549/commits/ecda7c5a5fed9f198984c4f6588607c067fe6e14